### PR TITLE
[AppConfig] Fix disposed stream access

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The focus of these changes is to fix a bug where the content stream returned by the protocol method has been disposed and the serializer attempts to access it.  Shifting to accessing the `BinaryData` content directly for deserialization.

# References and Resources

- [[BUG] ObjectDisposedException when calling GetConfigurationSettingAsync / SetConfigurationSettingAsync (#38219)](https://github.com/Azure/azure-sdk-for-net/issues/38219)

### CODE REVIEW

File: sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationMockTests.cs
Comment (jsquire) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/tests/ConfigurationMockTests.cs:20]:
```suggestion
```

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs
Comment (annelo-msft) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs:22]: One observation is that this creates a new MemoryStream under the hood.  Depending on how concerned we are with performance in this scenario, we could update the DeserializeSetting method to take a BinaryData.  If it's just using it to parse a JsonDocument, it should be a small change, and if you change the method signature, you might catch other instances of this bug while you're in there.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md
Comment (annelo-msft) [inline sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md:15]: Curious: do we know why we only hit it when we set a customer message handler and not in other cases?  It feels like that would be good to understand.

**Review by annelo-msft — APPROVED** [review]

File: sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md
Comment (jsquire) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md:15]: I believe this is because of the Core  HttpClientTransport and its PipelineResponse], which manipulates the content and replaces the content stream.   I haven't been able to trace it all the way through, though.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs
Comment (jsquire) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/src/ConfigurationClient_private.cs:22]: I've got some time to play with this one.   I'll poke around a bit and see how much churn moving to `BinaryData` would cause.

File: sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md
Comment (jsquire) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md:15]: This is related to the type of content set on the `HttpResponseMessage`.  

When `HttpMessage` is disposed, it disposes the `HttpResponseMessage` which, in turn, disposes its content.  In the case of `StringContent` and others derived from `HttpContent` the reference to the content stream is assumed to be owned by the content and disposed with it.

When forming the `PipelineResponse`, Core will read content stream and if it is a MemoryStream then it is considered buffered and kept without making a copy.  If it is another type of stream - such as a network stream - it is read into a MemoryStream.   

Because Core did not copy the stream read from `StringContent` set by the custom handler, it was disposed with `HttpMessage` despite appearing in `PipelineResponse` as a separate member.

Comment (annelo-msft) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md:15]: Oh killer.  Great investigation, @jsquire!  Would it make sense to update the buffering policy to handle this HttpContent type?  I know we've duplicated some of the buffering logic outside that policy, e.g. [here](https://github.com/Azure/azure-sdk-for-net/pull/35716/files#diff-9916010ee7060f5a4ec225b148221bfa381b0f235a4f9f4a740900f44439aaffR230) so we might need to fix it in multiple places and/or find a way to consolidate the "should I buffer" logic to a single place.  None of that needs to block this PR, though, but it would be good to file an issue to track that work if you haven't already.  Thank you!

Comment (pallavit) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md:15]: In theory this issue can happen in any SDK that has a similar pattern so shouldn't the right fix\onus of buffering\creating the copy be on the core pipeline pieces?

Comment (jsquire) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md:15]: The current generation doesn't have this problem because the response is processed in a scope where the `HttpMessage` is still alive.  Still not 100% sure whether this is a special case because it's old hand-written code being integrated into the DPG surface or if there's a legitimate pattern fix needed here.

I'd agree that we should consider revisiting the `PipelineResponse` approach and consider some alternate way to signal "this is buffered content owned by the pipeline" rather than just making assumptions based on the stream type.

That said, I don't have a deep familiarity with the pipeline implementation and would like to discuss with the Core team before we decide next steps.    If I'm correct, I believe this is a potential issue for all protocol methods.

Regardless, the AppConfig implementation is a hybrid approach due to the history.  We should still adjust the consumption pattern here.

Comment (pallavit) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md:15]: Perfect! My uber concern is if this is a legit DPG pattern we should move the fix up in Core than having each SDK fend for itself.

**Review by ShivangiReja — APPROVED** [review]

File: sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md
Comment (annelo-msft) [reply sdk/appconfiguration/Azure.Data.AppConfiguration/CHANGELOG.md:15]: > My uber concern is if this is a legit DPG pattern

I do think this is an issue of crossing the HLC/DPG streams, as Jesse indicated.  It would be good to have an issue filed in Core to track investigating whether we have a deeper problem here, if this is a use case we want to support.  I'm happy to let @jsquire drive it as he sees fit, though.

**Comment by jsquire** [comment]
Well, that rebase did not go well.   Closing this out to avoid the mass inclusion.

**Comment by jsquire** [comment]
Superseded by #38531, which fixes the bad rebase.

**Pull request closed without merging**
